compiler/luci-interpreter/src/kernels/Tanh.test.cpp

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (c) 2020 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "kernels/Tanh.h"
+#include "kernels/TestUtils.h"
+
+namespace luci_interpreter
+{
+namespace kernels
+{
+namespace
+{
+
+using namespace testing;
+
+TEST(TanhTest, Float)
+{
+  // TODO fix input & output ref data.
+
+  // TODO make a Shape checking of output_tensor.
+}
+
+// TODO Uint8
+// Need to Implement GetDequantizedOutput Function.
```

Comment:
We now have `quantize` and `dequantize` functions. Please refer tests using those functions, e.g., `MaxPool2D.test.cpp`.

Reply:
Thank you for your advice, @jinevening !
I'll follow your advice and fix these.

Reply:
@jinevening , could you please review code below?

<details>
<summary>Source Code</summary>
<div markdown="1">
<pre><code>
/*
 * Copyright (c) 2020 Samsung Electronics Co., Ltd. All Rights Reserved
 * Copyright 2017 The TensorFlow Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "kernels/Tanh.h"
#include "kernels/TestUtils.h"

namespace luci_interpreter
{
namespace kernels
{
namespace
{

using namespace testing;

TEST(TanhTest, Float)
{
  Shape input_shape{1, 2, 4, 1};
  std::vector<float> input_data{
      0, -6, 2,  4, //
      3, -2, 10, 1, //
  };
  Tensor input_tensor = makeInputTensor<DataType::FLOAT32>(input_shape, input_data);
  Tensor output_tensor = makeOutputTensor(DataType::FLOAT32);

  Tanh kernel(&input_tensor, &output_tensor);
  kernel.configure();
  kernel.execute();

  std::vector<float> ref_output_data{
      0, -0.9999877, 0.9640275, 0.999329,    //
      0.99505475, -0.9640275, 1, 0.7615941,  //
  };
  EXPECT_THAT(extractTensorData<float>(output_tensor),
              ElementsAreArray(ArrayFloatNear(ref_output_data)));
}

TEST(TanhTest, Uint8)
{
  float kMin = -1;
  float kMax = 127.f / 128.f;
  float kTanhTolerance = 0.014f;
  std::pair<float, int32_t> quant_param = quantizationParams<uint8_t>(kMin, kMax);
  std::vector<float> input_data{
    -8.0000000000, -7.8181818182, -7.6363636364, -7.4545454545, //
    -7.2727272727, -7.0909090909, -6.9090909091, -6.7272727273, //
    -6.5454545455, -6.3636363636, -6.1818181818, -6.0000000000, //
    -5.8181818182, -5.6363636364, -5.4545454545, -5.2727272727, //
    -5.0909090909, -4.9090909091, -4.7272727273, -4.5454545455, //
    -4.3636363636, -4.1818181818, -4.0000000000, -3.8181818182, //
    -3.6363636364, -3.4545454545, -3.2727272727, -3.0909090909, //
    -2.9090909091, -2.7272727273, -2.5454545455, -2.3636363636, //
    -2.1818181818, -2.0000000000, -1.8181818182, -1.6363636364, //
    -1.4545454545, -1.2727272727, -1.0909090909, -0.9090909091, //
    -0.7272727273, -0.5454545455, -0.3636363636, -0.1818181818, //
    0.0000000000,  0.1818181818,  0.3636363636,  0.5454545455, //
    0.7272727273,  0.9090909091,  1.0909090909,  1.2727272727, //
    1.4545454545,  1.6363636364,  1.8181818182,  2.0000000000, //
    2.1818181818,  2.3636363636,  2.5454545455,  2.7272727273, //
    2.9090909091,  3.0909090909,  3.2727272727,  3.4545454545, //
    3.6363636364,  3.8181818182,  4.0000000000,  4.1818181818, //
    4.3636363636,  4.5454545455,  4.7272727273,  4.9090909091, //
    5.0909090909,  5.2727272727,  5.4545454545,  5.6363636364, //
    5.8181818182,  6.0000000000,  6.1818181818,  6.3636363636, //
    6.5454545455,  6.7272727273,  6.9090909091,  7.0909090909, //
    7.2727272727,  7.4545454545,  7.6363636364,  7.8181818182, //
    8.0000000000, //
  };
  Tensor input_tensor{DataType::U8, {89}, {{quant_param.first * 8}, {quant_param.second * 8}}, ""};
  Tensor output_tensor = makeOutputTensor(DataType::U8, quant_param.first, quant_param.second);
  std::vector<uint8_t> quantize_input =
      quantize<uint8_t>(input_data, quant_param.first, quant_param.second);
  input_tensor.writeData(quantize_input.data(), quantize_input.size() * sizeof(uint8_t));

  Tanh kernel(&input_tensor, &output_tensor);
  kernel.configure();
  kernel.execute();

  std::vector<float> ref_output_data{
    -0.9999997749, -0.9999996762, -0.9999995342, -0.9999993300,
    -0.9999990361, -0.9999986134, -0.9999980053, -0.9999971306,
    -0.9999958722, -0.9999940619, -0.9999914578, -0.9999877117,
    -0.9999823226, -0.9999745703, -0.9999634183, -0.9999473758,
    -0.9999242982, -0.9998911009, -0.9998433469, -0.9997746542,
    -0.9996758446, -0.9995337191, -0.9993292997, -0.9990353053,
    -0.9986125310, -0.9980046622, -0.9971308601, -0.9958751909,
    -0.9940716137, -0.9914827859, -0.9877703933, -0.9824541388,
    -0.9748561217, -0.9640275801, -0.9486568273, -0.9269625051,
    -0.8965880154, -0.8545351057, -0.7972097087, -0.7206956332,
    -0.6213939966, -0.4971057414, -0.3484130125, -0.1798408185,
    0.0000000000,  0.1798408185,  0.3484130125,  0.4971057414,
    0.6213939966,  0.7206956332,  0.7972097087,  0.8545351057,
    0.8965880154,  0.9269625051,  0.9486568273,  0.9640275801,
    0.9748561217,  0.9824541388,  0.9877703933,  0.9914827859,
    0.9940716137,  0.9958751909,  0.9971308601,  0.9980046622,
    0.9986125310,  0.9990353053,  0.9993292997,  0.9995337191,
    0.9996758446,  0.9997746542,  0.9998433469,  0.9998911009,
    0.9999242982,  0.9999473758,  0.9999634183,  0.9999745703,
    0.9999823226,  0.9999877117,  0.9999914578,  0.9999940619,
    0.9999958722,  0.9999971306,  0.9999980053,  0.9999986134,
    0.9999990361,  0.9999993300,  0.9999995342,  0.9999996762,
    0.9999997749,
  };
  EXPECT_THAT(dequantize<uint8_t>(extractTensorData<uint8_t>(output_tensor),
                                  output_tensor.scale(), output_tensor.zero_point()),
              ElementsAreArray(ArrayFloatNear(ref_output_data, kTanhTolerance)));
}

} // namespace
} // namespace kernels
} // namespace luci_interpreter
</code></pre>
</div>
</details>

As a result of puttiing these test inputs in,
`./nncc build` passes without failure, but it fails with `./nncc test`.

<pre><code>
[==========] 90 tests from 40 test cases ran. (14 ms total)
[  PASSED  ] 89 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] TanhTest.Uint8
</code></pre>

I think I have an issue with my code.
Could you please review this code and let me know where to fix?

Source code I referred to is below:
<details>
<summary>Original Tensorflow Lite Source Code</summary>
<div markdown="1">
<pre><code>
TEST_P(TanhOpTest, TanhUint8) {
  const float kMin = -1;
  const float kMax = 127.f / 128.f;
  const float kTanhTolerance = 0.014f;
  QuantizedActivationsOpModel m(
      GetRegistration(), BuiltinOperator_TANH,
      /*input=*/{TensorType_UINT8, {89}, 8 * kMin, 8 * kMax},
      /*output=*/{TensorType_UINT8, {89}, kMin, kMax});
  // 64+16+8+1 elements, from -8 to 8.
  m.SetInput<uint8_t>(
      {-8.0000000000, -7.8181818182, -7.6363636364, -7.4545454545,
       -7.2727272727, -7.0909090909, -6.9090909091, -6.7272727273,
       -6.5454545455, -6.3636363636, -6.1818181818, -6.0000000000,
       -5.8181818182, -5.6363636364, -5.4545454545, -5.2727272727,
       -5.0909090909, -4.9090909091, -4.7272727273, -4.5454545455,
       -4.3636363636, -4.1818181818, -4.0000000000, -3.8181818182,
       -3.6363636364, -3.4545454545, -3.2727272727, -3.0909090909,
       -2.9090909091, -2.7272727273, -2.5454545455, -2.3636363636,
       -2.1818181818, -2.0000000000, -1.8181818182, -1.6363636364,
       -1.4545454545, -1.2727272727, -1.0909090909, -0.9090909091,
       -0.7272727273, -0.5454545455, -0.3636363636, -0.1818181818,
       0.0000000000,  0.1818181818,  0.3636363636,  0.5454545455,
       0.7272727273,  0.9090909091,  1.0909090909,  1.2727272727,
       1.4545454545,  1.6363636364,  1.8181818182,  2.0000000000,
       2.1818181818,  2.3636363636,  2.5454545455,  2.7272727273,
       2.9090909091,  3.0909090909,  3.2727272727,  3.4545454545,
       3.6363636364,  3.8181818182,  4.0000000000,  4.1818181818,
       4.3636363636,  4.5454545455,  4.7272727273,  4.9090909091,
       5.0909090909,  5.2727272727,  5.4545454545,  5.6363636364,
       5.8181818182,  6.0000000000,  6.1818181818,  6.3636363636,
       6.5454545455,  6.7272727273,  6.9090909091,  7.0909090909,
       7.2727272727,  7.4545454545,  7.6363636364,  7.8181818182,
       8.0000000000});
  m.Invoke();
  EXPECT_THAT(m.GetDequantizedOutput<uint8_t>(),
              ElementsAreArray(ArrayFloatNear(
                  {-0.9999997749, -0.9999996762, -0.9999995342, -0.9999993300,
                   -0.9999990361, -0.9999986134, -0.9999980053, -0.9999971306,
                   -0.9999958722, -0.9999940619, -0.9999914578, -0.9999877117,
                   -0.9999823226, -0.9999745703, -0.9999634183, -0.9999473758,
                   -0.9999242982, -0.9998911009, -0.9998433469, -0.9997746542,
                   -0.9996758446, -0.9995337191, -0.9993292997, -0.9990353053,
                   -0.9986125310, -0.9980046622, -0.9971308601, -0.9958751909,
                   -0.9940716137, -0.9914827859, -0.9877703933, -0.9824541388,
                   -0.9748561217, -0.9640275801, -0.9486568273, -0.9269625051,
                   -0.8965880154, -0.8545351057, -0.7972097087, -0.7206956332,
                   -0.6213939966, -0.4971057414, -0.3484130125, -0.1798408185,
                   0.0000000000,  0.1798408185,  0.3484130125,  0.4971057414,
                   0.6213939966,  0.7206956332,  0.7972097087,  0.8545351057,
                   0.8965880154,  0.9269625051,  0.9486568273,  0.9640275801,
                   0.9748561217,  0.9824541388,  0.9877703933,  0.9914827859,
                   0.9940716137,  0.9958751909,  0.9971308601,  0.9980046622,
                   0.9986125310,  0.9990353053,  0.9993292997,  0.9995337191,
                   0.9996758446,  0.9997746542,  0.9998433469,  0.9998911009,
                   0.9999242982,  0.9999473758,  0.9999634183,  0.9999745703,
                   0.9999823226,  0.9999877117,  0.9999914578,  0.9999940619,
                   0.9999958722,  0.9999971306,  0.9999980053,  0.9999986134,
                   0.9999990361,  0.9999993300,  0.9999995342,  0.9999996762,
                   0.9999997749},
                  kTanhTolerance)));
}
</code></pre>
</div>
</details>